contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,329 @@ const (
+domain that the selected node is part of. The domain depends on the volume
+plugin. Zonal volume plugins will create the volume in the zone where the
+selected node is in. The local volume plugin will create the volume on the
+selected node.
+
+### End to End Zonal Example
+This is an example of the most common use case for provisioning zonal volumes.
+For this use case, the user's specs are unchanged. Only one change
+to the StorageClass is needed to enable delayed volume binding.
+
+1. Admin sets up StorageClass, setting up delayed volume binding.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: standard
+provisioner: kubernetes.io/gce-pd
+volumeBindingMode: WaitForFirstConsumer
+parameters:
+  type: pd-standard
+```
+2. Admin launches provisioner.  For in-tree plugins, nothing needs to be done.
+3. User creates PVC. Nothing changes in the spec, although now the PVC won't be
+   immediately bound.
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: my-pvc
+spec:
+  storageClassName: standard
+  accessModes:
+  - ReadWriteOnce
+  resources:
+    requests:
+      storage: 100Gi
+```
+4. User creates Pod. Nothing changes in the spec.
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-pod
+spec:
+  containers:
+  ...
+  volumes:
+  - name: my-vol
+    persistentVolumeClaim:
+      claimName: my-pvc
+```
+5. Scheduler picks a node that can satisfy the Pod and passes it to the
+   provisioner.
+6. Provisioner dynamically provisions a PV that can be accessed from
+   that node.
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: standard
+  gcePersistentDisk:
+    diskName: my-disk
+    fsType: ext4
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: failure-domain.beta.kubernetes.io/zone
+          operator: In
+          values:
+          - us-central1-a
+```
+7. Pod gets scheduled to the node.
+
+
+### Restricting Topology
+For the common use case, volumes will be provisioned in whatever topology domain
+the scheduler has decided is best to run the workload. Users may impose further
+restrictions by setting label/node selectors, and pod affinity/anti-affinity
+policies on their Pods. All those policies will be taken into account when
+dynamically provisioning a volume.
+
+While less common, administrators may want to further restrict what topology
+domains are available to a StorageClass. To support these administrator
+policies, a VolumeProvisioningTopology field can also be specified in the
+StorageClass to restrict the allowed topology domains for dynamic provisioning.
+This is not expected to be a common use case, and there are some caveats,
+described below.
+
+```
+type StorageClass struct {
+    ...
+
+    AllowedTopology *VolumeProvisioningTopology
+}
+
+type VolumeProvisioningTopology struct {
+    Required map[string]VolumeTopologyValues
```

Comment:
@saad-ali found a use case that this cannot represent:
An administrator wants to restrict provisioning to either "zone A, rack 1" or "zone B, rack 1".

To fix this, we could go back to using a restricted node selector (only support the IN operator), or follow the CSI [proposed](https://github.com/container-storage-interface/spec/pull/188#issuecomment-366425374) representation which sort of does the reverse of the node selector: OR, then AND.

A restricted NodeSelector would be more familiar to Kubernetes users, so I'm leaning towards that approach.   As long as we restrict the operators to just IN, then it can be translated internally into the CSI representation.

Reply:
I'm flip flopping again.  A node selector specifies a subset off of a set of labels on the node object.  But in this case, we are not selecting from a set of labels on an existing object.  In this case, we want to limit the topology of a PV object that will be created in the future.  So I think following the CSI representation as a list of maps makes more sense, rather then specifying a selector on nothing.

Reply:
+1

Reply:
+1. Another semantic difference is: VolumeTopology is used to select **a set of nodes** at which a PV is available, whereas NodeSelector is used to select **a single node**.

Reply:
> NodeSelector is used to select a single node

that's a little surprising... selectors typically select a set of things

Reply:
@liggitt you are right, I was using a wrong frame of thinking, so I take that back.

Here's another point, similar to what @msau42  said, that helped me understand the difference: ultimately the scheduler needs a set of nodes to choose from; a NodeSelector **directly** specifies that set, whereas VolumeTopology **indirectly** specifies it by providing a super set which is then refined further by the volume plugin. Please correct me if this is the wrong understanding.